File: proxyvirus.java

interface Archivo {
    void ejecutar();
}
class Virus implements Archivo {
    private String nombre;

    public Virus(String nombre) {
        this.nombre = nombre;
    }

    @Override
    public void ejecutar() {
        System.out.println(" Ejecutando virus: " + nombre + " ");
    }
}

// Proxy que controla el acceso
class ProxyVirus implements Archivo {
    private Virus virus;
    private String usuario;
    private String nombre;

    public ProxyVirus(String nombre, String usuario) {
        this.nombre = nombre;
        this.usuario = usuario;
    }

    @Override
    public void ejecutar() {
        if (usuario.equals("admin")) {
            if (virus == null) {
                virus = new Virus(nombre);
            }
            virus.ejecutar();
        } else {
            System.out.println(" Acceso denegado al virus " + nombre + " para el usuario " + usuario);
        }
    }
}
public class proxyvirus {
    public static void main(String[] args) {
        Archivo virus1 = new ProxyVirus("Troyano.exe", "admin");
        virus1.ejecutar();

        Archivo virus2 = new ProxyVirus("Troyano.exe", "invitado");
        virus2.ejecutar();
    }
    
}

File: proxyvirus.py
from abc import ABC, abstractmethod

class Archivo(ABC):
    @abstractmethod
    def ejecutar(self):
        pass

class Virus(Archivo):
    def __init__(self, nombre):
        self.nombre = nombre
    
    def ejecutar(self):
        print(f" Ejecutando el virus: {self.nombre}... ")

class ProxyVirus(Archivo):
    def __init__(self, nombre, usuario):
        self.nombre = nombre
        self.usuario = usuario
        self.virus_real = None
    
    def ejecutar(self):
        if self.usuario == "admin":
            if not self.virus_real:
                self.virus_real = Virus(self.nombre)
            self.virus_real.ejecutar()
        else: 
            print(f" Acceso denegado: El usuario '{self.usuario}' no tiene permisos para ejecutar {self.nombre}")

# Simulación de acceso al virus
archivo1 = ProxyVirus("troyano.exe", "admin")
archivo1.ejecutar()

archivo2 = ProxyVirus("troyano.exe", "invitado")
archivo2.ejecutar()

